opentreemap/treemap/templatetags/species_thumbprint.py

```diff
@@ -0,0 +1,23 @@
+from django import template
+
+register = template.Library()
+
+last_request_and_thumbprint = (None, None)
+
+# Some pages contain many species autocompletes. To avoid re-querying
+# we hang on to the last request and thumbprint. This isn't thread-safe,
+# but the the worst that could happen with a data race is an extra query.
```

Comment:
I was thinking about two threads making interleaved calls to `species_thumbprint`. Neither would get the wrong answer, but they would keep invalidating each other's cache.

This might solve it:

```python
import threading

_thread_local = threading.local()

def species_thumbprint(request):
    last_request, last_thumbprint = _thread_local.getattr(
        last_request_and_thumbprint, (None, None))

    if request == last_request:
        thumbprint = last_thumbprint
    else:
        thumbprint = request.instance.species_thumbprint
        _thread_local.last_request_and_thumbprint = (request, thumbprint)

    return thumbprint
```